beets/util/artresizer.py

```diff
@@ -61,327 +57,409 @@ def temp_file_for(path):
         return bytestring_path(f.name)
 
 
-def pil_resize(artresizer, maxwidth, path_in, path_out=None, quality=0,
-               max_filesize=0):
-    """Resize using Python Imaging Library (PIL).  Return the output path
-    of resized image.
-    """
-    path_out = path_out or temp_file_for(path_in)
-    from PIL import Image
+class LocalBackendNotAvailableError(Exception):
+    pass
 
-    log.debug('artresizer: PIL resizing {0} to {1}',
-              displayable_path(path_in), displayable_path(path_out))
 
-    try:
-        im = Image.open(syspath(path_in))
-        size = maxwidth, maxwidth
-        im.thumbnail(size, Image.ANTIALIAS)
+_NOT_AVAILABLE = object()
 
-        if quality == 0:
-            # Use PIL's default quality.
-            quality = -1
 
-        # progressive=False only affects JPEGs and is the default,
-        # but we include it here for explicitness.
-        im.save(py3_path(path_out), quality=quality, progressive=False)
+class LocalBackend:
+    @classmethod
+    def available(cls):
+        try:
+            cls.version()
+            return True
+        except LocalBackendNotAvailableError:
+            return False
 
-        if max_filesize > 0:
-            # If maximum filesize is set, we attempt to lower the quality of
-            # jpeg conversion by a proportional amount, up to 3 attempts
-            # First, set the maximum quality to either provided, or 95
-            if quality > 0:
-                lower_qual = quality
-            else:
-                lower_qual = 95
-            for i in range(5):
-                # 5 attempts is an abitrary choice
-                filesize = os.stat(syspath(path_out)).st_size
-                log.debug("PIL Pass {0} : Output size: {1}B", i, filesize)
-                if filesize <= max_filesize:
-                    return path_out
-                # The relationship between filesize & quality will be
-                # image dependent.
-                lower_qual -= 10
-                # Restrict quality dropping below 10
-                if lower_qual < 10:
-                    lower_qual = 10
-                # Use optimize flag to improve filesize decrease
-                im.save(py3_path(path_out), quality=lower_qual,
-                        optimize=True, progressive=False)
-            log.warning("PIL Failed to resize file to below {0}B",
-                        max_filesize)
-            return path_out
 
-        else:
-            return path_out
-    except OSError:
-        log.error("PIL cannot create thumbnail for '{0}'",
-                  displayable_path(path_in))
-        return path_in
+class IMBackend(LocalBackend):
+    NAME = "ImageMagick"
+    _version = None
+    _legacy = None
```

Comment:
Maybe worth a quick comment to indicate what this field means—i.e., whether we're using the `magick` binary?